Automattic/newspack-ads
# feat(header-bidding): multiple gam orders

Implement multiple GAM orders support for header bidding in order to meet the different revenue shares each bidder agreement requires.

<img width="795" alt="image" src="https://user-images.githubusercontent.com/820752/154141462-45ec595f-77ec-4433-aea8-76902da3f8ce.png">

<img width="798" alt="image" src="https://user-images.githubusercontent.com/820752/154141522-53434b3b-aa51-454a-b587-259f360c27d8.png">

## How to test this PR

⚠️ The default price granularity makes the creation of orders take up to 15 minutes. For testing purposes, if you'd like to review with less granular buckets, add `price_granularity_key: 'low'` to line `119` of `src/wizard-settings/header-bidding-gam/order.js` so it generates a smaller set of line items.

1. Check out this branch and https://github.com/Automattic/newspack-plugin/pull/1517
2. Make sure your instance is properly connected to GAM
3. Visit Newspack Ads -> Settings
4. Configure all bidders with random values and make sure they are enabled
5. Click on Manage Orders or Create Order (the message varies according to existing orders)
6. If you have existing orders, observe the list of orders available and click "Edit" on any of them
7. If there are any found issues (missing local configuration), click on Fix issues
8. After the issues are fixed, change the shared revenue to any percentage and select a couple of bidders
9. Update the order and visit an order line item GAM's dashboard to confirm the percentage has been subtracted to the unit value and the targeting is set to the selected bidders:

<img width="587" alt="Captura de Tela 2022-02-16 às 14 36 39" src="https://user-images.githubusercontent.com/820752/154323369-b074f385-abe3-4d38-90a7-5640abd96f59.png">

<img width="1014" alt="Captura de Tela 2022-02-16 às 14 37 12" src="https://user-images.githubusercontent.com/820752/154323393-922c835f-fd1a-4df2-a9ca-fc0b0436a36b.png">

### Multiple orders

<img width="672" alt="image" src="https://user-images.githubusercontent.com/820752/154324002-27834ef8-a463-4406-8182-30600a1a81d7.png">

1. Create a new order with a random revenue share and different bidders and make sure you have a duplicate bidder from the previous order
2. Observe the warning message stating that there are 2 orders targeting that bidder
3. Edit the order to remove the duplicate bidder and change the revenue share to another random value
4. Save the changes and confirm the warning message is gone and the updated values are matched on GAM's dashboard

### Archive order

1. While on the "Manage Orders" Modal, click to Archive an order
2. Confirm the order is archived on GAM
3. Archive all other orders and confirm your modal falls back to create a new one.
